include/fiction/algorithms/simulation/sidb/clustercomplete.hpp

```diff
@@ -0,0 +1,479 @@
+//
+// Created by Willem Lambooy on 29.02.2024.
+//
+
+#ifndef FICTION_CLUSTERCOMPLETE_HPP
+#define FICTION_CLUSTERCOMPLETE_HPP
+
+#if (FICTION_ALGLIB_ENABLED)
+
+#include "fiction/algorithms/simulation/sidb/ground_state_space.hpp"
+#include "fiction/algorithms/simulation/sidb/sidb_simulation_result.hpp"
+#include "fiction/technology/charge_distribution_surface.hpp"
+#include "fiction/technology/physical_constants.hpp"
+#include "fiction/technology/sidb_charge_state.hpp"
+#include "fiction/technology/sidb_cluster_hierarchy.hpp"
+#include "fiction/utils/hash.hpp"
+
+#include <mockturtle/utils/stopwatch.hpp>
+
+#include <algorithm>
+#include <array>
+#include <cstdint>
+#include <iterator>
+#include <mutex>
+#include <thread>
+#include <utility>
+#include <vector>
+
+namespace fiction
+{
+
+/**
+ * The struct containing the parameters both passed on to pre-simulator Ground State Space, and used during simulation.
+ */
+struct clustercomplete_params
+{
+    /**
+     * Physical simulation parameters.
+     */
+    sidb_simulation_parameters physical_parameters{};
+    /**
+     * This specifies the maximum cluster size for which Ground State Space will solve an NP-complete sub-problem
+     * exhaustively. The sets of SiDBs that witness local population stability for each respective charge state may be
+     * partitioned into disjoint sets such that the number of required witnesses for each respective charge state is
+     * satisfied. If no such partition exists, the multiset charge configuration associated with the requirements may be
+     * rejected.
+     */
+    uint64_t validity_witness_partitioning_max_cluster_size_gss = 6;
+    /**
+     * Number of threads to make available to ClusterComplete for the simulation stage.
+     */
+    uint64_t available_threads = std::thread::hardware_concurrency();
+    /**
+     * Report the *Ground State Space* statistics to standard output.
+     */
+    bool report_gss_stats = false;
```

Comment:
I would recommend adding a ``stats`` structure. There you can also define a report function etc. Have a look at the critical temperature.